File: sorts_searches/mergesort.js
// Time complelxity: O(n log(n))
// Space complexity: O(n)

function merge(left, right) {
  let res = [];
  while (right.length > 0 && left.length > 0) {
    if (right[0] <= left[0]) {
      res.push(right[0]);
      right.splice(0,1);
    } else {
      res.push(left[0]);
      left.splice(0,1);
    }
  }
  return res.concat(left).concat(right);
}


function mergesort(arr) {
  if (arr.length <= 1) return arr;

  const midx = Math.floor(arr.length / 2);
  let left = mergesort(arr.slice(0, midx));
  let right = mergesort(arr.slice(midx));

  return merge(left, right);

}

console.log(mergesort([3,2,517,9,7,2]));

File: essential_sorts_searches/ruby/sorts/mergesort.rb

def mergesort(arr)
  return arr if arr.length <= 1

  mid = arr.length / 2
  left = arr.take(mid)
  right = arr.drop(mid)

  merge(mergesort(left), mergesort(right))
end

def merge(left, right)
  merged = []
  i, j = 0, 0

  while i < left.length && j < right.length
    if left[i] < right[j]
      merged << left[i]
      i += 1
    else
      merged << right[j]
      j += 1
    end
  end

  merged + left.drop(i) + right.drop(j)
end

p mergesort([1,4,6,3,7,3])
